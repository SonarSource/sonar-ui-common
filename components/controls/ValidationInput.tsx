/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import AlertErrorIcon from '../icons/AlertErrorIcon';
import AlertSuccessIcon from '../icons/AlertSuccessIcon';
import HelpTooltip from './HelpTooltip';

interface Props {
  description?: string;
  children: React.ReactNode;
  className?: string;
  error: string | undefined;
  help?: string;
  id: string;
  isInvalid: boolean;
  isValid: boolean;
  label: React.ReactNode;
  required?: boolean;
}

export default function ValidationInput(props: Props) {
  const hasError = props.isInvalid && props.error !== undefined;
  return (
    <div className={props.className}>
      <label htmlFor={props.id}>
        <span className="text-middle">
          <strong>{props.label}</strong>
          {props.required && <em className="mandatory">*</em>}
        </span>
        {props.help && <HelpTooltip className="spacer-left" overlay={props.help} />}
      </label>
      <div className="little-spacer-top spacer-bottom">
        {props.children}
        {props.isInvalid && <AlertErrorIcon className="spacer-left text-middle" />}
        {hasError && (
          <span className="little-spacer-left text-danger text-middle">{props.error}</span>
        )}
        {props.isValid && <AlertSuccessIcon className="spacer-left text-middle" />}
      </div>
      {props.description && <div className="note abs-width-400">{props.description}</div>}
    </div>
  );
}
