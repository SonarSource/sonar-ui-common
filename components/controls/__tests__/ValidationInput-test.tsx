/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import ValidationInput from '../ValidationInput';

it('should render', () => {
  expect(
    shallow(
      <ValidationInput
        description="My description"
        error={undefined}
        help="Help message"
        id="field-id"
        isInvalid={false}
        isValid={false}
        label="Field label"
        required={true}>
        <div />
      </ValidationInput>
    )
  ).toMatchSnapshot();
});

it('should render with error', () => {
  expect(
    shallow(
      <ValidationInput
        description={<div>My description</div>}
        error="Field error message"
        id="field-id"
        isInvalid={true}
        isValid={false}
        label="Field label">
        <div />
      </ValidationInput>
    )
  ).toMatchSnapshot();
});

it('should render when valid', () => {
  expect(
    shallow(
      <ValidationInput
        description="My description"
        error={undefined}
        id="field-id"
        isInvalid={false}
        isValid={true}
        label="Field label"
        required={true}>
        <div />
      </ValidationInput>
    )
  ).toMatchSnapshot();
});
