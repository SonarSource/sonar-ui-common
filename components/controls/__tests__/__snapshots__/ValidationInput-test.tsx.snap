// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div>
  <label
    htmlFor="field-id"
  >
    <span
      className="text-middle"
    >
      <strong>
        Field label
      </strong>
      <em
        className="mandatory"
      >
        *
      </em>
    </span>
    <HelpTooltip
      className="spacer-left"
      overlay="Help message"
    />
  </label>
  <div
    className="little-spacer-top spacer-bottom"
  >
    <div />
  </div>
  <div
    className="note abs-width-400"
  >
    My description
  </div>
</div>
`;

exports[`should render when valid 1`] = `
<div>
  <label
    htmlFor="field-id"
  >
    <span
      className="text-middle"
    >
      <strong>
        Field label
      </strong>
      <em
        className="mandatory"
      >
        *
      </em>
    </span>
  </label>
  <div
    className="little-spacer-top spacer-bottom"
  >
    <div />
    <AlertSuccessIcon
      className="spacer-left text-middle"
    />
  </div>
  <div
    className="note abs-width-400"
  >
    My description
  </div>
</div>
`;

exports[`should render with error 1`] = `
<div>
  <label
    htmlFor="field-id"
  >
    <span
      className="text-middle"
    >
      <strong>
        Field label
      </strong>
    </span>
  </label>
  <div
    className="little-spacer-top spacer-bottom"
  >
    <div />
    <AlertErrorIcon
      className="spacer-left text-middle"
    />
    <span
      className="little-spacer-left text-danger text-middle"
    >
      Field error message
    </span>
  </div>
  <div
    className="note abs-width-400"
  >
    <div>
      My description
    </div>
  </div>
</div>
`;
